lib/shared/option-utils.js

```diff
@@ -0,0 +1,50 @@
+/**
+ * @fileoverview Utilities to operate on arrays.
+ * @author Josh Goldberg
+ */
+
+"use strict";
+
+/**
+ * Determines whether any of input's properties are different
+ * from values that already exist in original.
+ * @template T
+ * @param {Partial<T>} input New value.
+ * @param {T} original Original value.
+ * @returns {boolean} Whether input includes an explicit difference.
+ */
+function containsDifferentProperty(input, original) {
+    if (input === original) {
+        return false;
+    }
+
+    if (Array.isArray(input)) {
+        return (
+            !Array.isArray(original) ||
+            input.length !== original.length ||
+            input.some((value, i) =>
+                containsDifferentProperty(value, original[i]))
+        );
+    }
+
+    if (typeof input !== typeof original) {
+        return true;
+    }
+
+    if (typeof input === "object") {
+        const inputKeys = Object.keys(input);
+        const originalKeys = Object.keys(original);
+
+        return inputKeys.length !== originalKeys.length || inputKeys.some(
+            inputKey =>
+                !Object.hasOwn(original, inputKey) ||
+                containsDifferentProperty(input[inputKey], original[inputKey])
+        );
+    }
```

Comment:
This could crash on a `null` value. For example:

```js
// eslint.config.js
export default [
    {
        linterOptions: {
            reportUnusedInlineConfigs: "error"
        },
        plugins: {
            "my-plugin": {
                rules: {
                    "my-rule": {
                        meta: {
                            schema: false
                        },
                        create() {
                            return {};
                        }
                    }
                }
            }
        },
        rules: {
            "my-plugin/my-rule": ["error", null]
        }
    }
];

```

```js
// foo.js

/* eslint my-plugin/my-rule: ["error", {}] */

```

```
  3:1  error  Inline configuration for rule "my-plugin/my-rule" is invalid:
        Cannot convert undefined or null to object
  my-plugin/my-rule
```